File: backend/src/main/java/ibee/webapp/todo_app/core/dto/person/skills/hard/PersonDegreeDto.java
package ibee.webapp.todo_app.core.dto.person.skills.hard;

import java.time.LocalDate;

import ibee.webapp.todo_app.core.dto.EducationInstitutionDto;
import ibee.webapp.todo_app.core.dto.person.referenceIds.skill.hard.PersonDegreeDtoId;
import ibee.webapp.todo_app.core.dto.skills.hard.DegreeDto;
import jakarta.validation.constraints.DecimalMax;
import jakarta.validation.constraints.DecimalMin;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.PastOrPresent;

public record PersonDegreeDto(

    PersonDegreeDtoId id,

    @NotNull
    DegreeDto degree,

    @NotNull
    EducationInstitutionDto educationInstitution,

    /*
        start date of this degree
    */
    @NotNull
    @PastOrPresent
    LocalDate startDate,

    /*
        end Date of this degree
    */
    @PastOrPresent
    LocalDate endDate,

    /*
        how much persent are finisched of this degree
    */
    @NotNull
    @DecimalMin(value = "0.0")
    @DecimalMax(value = "100.0")
    double progressInPercent


) {

}

File: frontend/src/app/types/dto/person/related/skill/hard/person-degree.dto.ts
import { DegreeDto } from '../../../../skill/hard/degree.dto';
import { EducationInstitutionDto } from '../../../../education-institution.dto';
import { PersonDegreeDtoId } from '../../reference/skill/person-degree-dto-id';


export interface PersonDegreeDto {

    id?: PersonDegreeDtoId;

    degree: DegreeDto;

    educationInstitution: EducationInstitutionDto;

    startDate: Date;

    endDate?: Date;

    progressInPercent: number;

}